### DISCUSSION TITLE
MongoDB Atlas error: OP_QUERY is no longer supported

### DISCUSSION THREAD

**Discussion by bjun626** [emqx/emqx#11428] [Q&A] [answered]
I'm testing auth integration using Atlas M0 with emqx/emqx:5.1.4 docker. I'm getting the error of 

`[warning] msg: emqx_connector_mongo_health_check, mfa: emqx_mongodb:on_get_status/2, line: 304, instance_id: <<"emqx_authz_mongodb:8">>, reason: {error,{bad_query,#{<<"$err">> => <<"OP_QUERY is no longer supported. The client driver may require an upgrade. For more details see https://dochub.mongodb.org/core/legacy-opcode-removal">>,<<"code">> => 5739101,<<"ok">> => 0.0}}}, status: failed`

Based on #9085 it seems the MongoDB 6 is supported but this error seems like an outdated driver in EMQX or I have misconfigured the integration.

Below attached my  `cluster.hocon`
<details>

```hocon
...
authorization {
  cache {
    enable = true
    max_size = 32
    ttl = 1m
  }
  deny_action = ignore
  no_match = allow
  sources = [
    {
      auth_source = admin
      collection = users
      database = mqtt
      filter {username = "${username}"}
      mongo_type = rs
      password = MONGODB_PASSWORD
      pool_size = 8
      r_mode = master
      replica_set_name = atlas-XXXXXX-shard-X
      servers = XXXXXX.myenxms.mongodb.net
      srv_record = true
      ssl {enable = true, verify = verify_none}
      topology {connect_timeout_ms = 20s}
      type = mongodb
      username = MONGODB_USERNAME
      w_mode = unsafe
    },
    {
      enable = true
      path = "data/authz/acl.conf"
      type = file
    }
  ]
}
...
```

![Screenshot 2023-08-11 003539](https://github.com/emqx/emqx/assets/34598848/d2331413-a27d-47a3-94b3-238d116959a2)

</details>

Please let me know if I missed anything. Thanks.

Version:
- MongoDB: Atlas M0 MongoDB 6.0.8 Enterprise
- EMQX: Docker emqx/emqx:5.1.4

**Comment by thalesmg** [comment, marked as answer]
Hi @bjun626 , thanks for your report.

The mongodb driver attempts to autodetect the protocol version used by the server, and it could have failed to do so correctly.

Could you try running the following command on your EMQX container to see if the behavior is fixed?

```sh
emqx eval 'application:set_env(mongodb, use_legacy_protocol, false).'
```

**Reply by bjun626** [reply]
The command works!!! 

But I'm not sure how to set it so that the env is persistent

**Reply by thalesmg** [reply]
That's great to hear!

It means that the automatic detection indeed failed in your case, for some reason.  Unfortunately, at the present moment, it's not possible to configure that persistently.  We'll have to make that a configurable option.

**Reply by thalesmg** [reply]
It should land within the next couple releases: https://github.com/emqx/emqx/pull/11429

**Reply by bjun626** [reply]
Awesome. Thanks for the swift fix.
